### PR TITLE
fix: encapsulate truncate_to_token_budget behind stable public API

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

- Changed `truncate_to_token_budget` in `truncate.rs` from `pub fn` to `pub(crate) fn` so the internal implementation no longer leaks through the module path
- Added a public wrapper function in `lib.rs` that delegates to the internal implementation with full doc comments, examples, and the empty-string edge case documented
- Removed the `pub use` re-export — consumers now call the wrapper, which is a proper stable API surface

## Test plan

- [x] `cargo build --workspace` — CLI crate resolves import through new wrapper
- [x] `cargo test --workspace` — 279/279 tests pass
- [x] `cargo test --doc -p rskim-core` — 11/11 doc tests pass (including new wrapper)
- [x] `cargo bench --no-run` — benchmark crate resolves import
- [x] `cargo clippy --workspace -- -D warnings` — zero warnings
- [x] `cargo doc --no-deps -p rskim-core` — function appears under Functions (not as re-export)
- [x] Budget invariant stress test: output tokens ≤ budget for 24/24 file × budget combos
- [x] Edge cases verified: budget 0/1/2-5/exact/10M/10M+1, empty file, stdin, glob, cache keys

### CODE REVIEW

File: crates/rskim-core/src/lib.rs [hunk 1]
```diff
@@ -238,6 +235,59 @@ pub fn transform_detailed(source: &str, language: Language, mode: Mode) -> Resul
...
+///
+/// ```no_run
+/// use rskim_core::{truncate_to_token_budget, Language};
+///
+/// let output = "line 1\nline 2\nline 3\nline 4\nline 5";
+/// let word_count = |s: &str| -> usize { s.split_whitespace().count() };
+/// let truncated = truncate_to_token_budget(output, Language::TypeScript, 5, word_count, None)?;
+/// # Ok::<(), rskim_core::SkimError>(())
+/// ```
+pub fn truncate_to_token_budget<F>(
```
Comment (dean0x) [inline crates/rskim-core/src/lib.rs]: **Consider adding `#[inline]` to thin delegation wrapper**

This public wrapper function is a pure pass-through delegation with zero logic -- it merely forwards all five arguments to the internal implementation. While the compiler will almost certainly inline it at release optimization levels, adding `#[inline]` would:

1. Make the intent explicit for future maintainers
2. Ensure inlining across crate boundaries (from `rskim` CLI into `rskim-core`)
3. Align with the "zero-cost abstraction" principle stated in the crate module docs (line 32)

The impact in practice is negligible (the compiler handles this anyway via LLVM LTO), but explicit annotation is a best practice for thin delegation wrappers.

**Suggested fix:**

```rust
#[inline]
pub fn truncate_to_token_budget<F>(
    text: &str,
    language: Language,
    token_budget: usize,
    count_tokens: F,
    known_token_count: Option<usize>,
) -> Result<String>
where
    F: Fn(&str) -> usize,
{
    // ...
}
```

**Review source**: consistency.md (MEDIUM priority - Should Fix)

File: crates/rskim-core/src/lib.rs [hunk 2]
```diff
@@ -238,6 +235,59 @@ pub fn transform_detailed(source: &str, language: Language, mode: Mode) -> Resul
...
+/// * `token_budget` - Maximum number of tokens allowed
+/// * `count_tokens` - Closure that counts tokens in a string slice
+/// * `known_token_count` - Pre-computed token count of `text`, if available.
+///   When `Some(count)`, skips the initial full-text tokenization.
+///   Pass `None` when the count is unknown.
+///
+/// # Returns
+/// Text fitting within the token budget, with omission marker if truncated.
+/// If the budget is smaller than the omission marker itself (~5-7 tokens),
+/// an empty string is returned rather than violating the budget invariant.
```
Comment (dean0x) [inline crates/rskim-core/src/lib.rs]: **Missing `# Errors` documentation section**

The rustdoc comment on `truncate_to_token_budget` follows the structure: Arguments → Returns → Examples. However, since this function returns `Result<String>`, the documentation should include an `# Errors` section to be consistent with other public functions in this file (like `transform` and `transform_with_config`).

Even though the function currently does not produce errors in practice, documenting error behavior improves API completeness and helps maintainers understand the contract.

**Suggested fix:** Add an `# Errors` section before the `# Examples` section:

```rust
/// # Errors
///
/// Returns `Err` if the internal truncation logic encounters an unexpected state.
/// In practice this function does not currently produce errors, but returns
/// `Result` for forward-compatibility with the rest of the API.
```

**Review source**: consistency.md (LOW priority - documentation gap)

File: crates/rskim-core/src/lib.rs (hunk 1 above)
Comment (dean0x) [inline crates/rskim-core/src/lib.rs]: ## Summary of Code Review Findings

**PR**: #30 - Encapsulate truncate_to_token_budget behind stable public API
**Branch**: fix/public-api-truncate-visibility → main
**Review Date**: 2026-03-15

### Overall Assessment
This is a clean API encapsulation refactor with no logic changes. The PR replaces a leaky `pub use` re-export with a proper public wrapper function, improving API stability and reducing internal surface. All 151 tests pass. Zero blocking security, architecture, or performance issues found.

### Approval Status
✅ **APPROVED** (4 inline suggestions, no blockers)

---

### Detailed Findings by Category

#### Architecture (Score: 9/10)
**Inline comments**: 1 (delegation depth consistency)
**Summary**: The new wrapper pattern is architecturally sound. It decouples the public API surface from internal module structure, preventing downstream breakage if `transform::truncate` is reorganized. This aligns with the existing patterns for `detect_language()` and `transform()`.

#### Performance (Score: 9/10)
**Status**: No regressions. The wrapper is a zero-cost abstraction that the compiler will inline in release builds.

#### Consistency (Score: 9/10)
**Inline comments**: 2 (`#[inline]` annotation, `# Errors` doc section)
**Summary**: Minor documentation and annotation suggestions. The wrapper follows existing patterns; these are quality-of-life improvements.

#### Tests (Score: 8/10)
**Inline comments**: 2 (missing wrapper test, runnable doc-test)
**Summary**: The internal function has excellent unit coverage (15 tests). The gap is at the library integration level -- no test exercises the public wrapper through the crate-level import path. The doc-test uses `no_run`, which only validates compilation.

#### Regression (Score: 10/10)
**Status**: Zero regression risk. Public export path is preserved, signature unchanged, all consumers compile without modification.

#### Security (Score: 9/10)
**Status**: No blocking security issues. The visibility restriction from `pub` to `pub(crate)` is itself a security improvement (reduced public API surface). Two pre-existing LOW observations documented but do not warrant blocking.

#### Rust Quality (Score: 9/10)
**Status**: Clean encapsulation pattern. All tests pass, clippy clean, doc-tests compile.

#### Complexity (Score: 9/10)
**Status**: The wrapper has cyclomatic complexity of 1 (trivial pass-through). The PR actually *reduces* net complexity by eliminating the `pub use` re-export.

---

### Recommended Next Steps

**Before merge** (if desired):
1. Add an integration test that calls `rskim_core::truncate_to_token_budget` directly via the public crate path
2. Convert the doc-test from `no_run` to runnable
3. Optionally add `#[inline]` annotation
4. Optionally add `# Errors` doc section

**Post-merge** (lower priority):
- Consider monitoring `lib.rs` file length as more public API functions are added (~380 lines currently)

---

### Review Sources
- security.md (0 blocking, 2 pre-existing LOW)
- architecture.md (0 blocking, 1 MEDIUM should-fix)
- performance.md (0 blocking, 0 should-fix, 2 pre-existing LOW)
- complexity.md (0 blocking, 1 LOW should-fix, 2 pre-existing)
- consistency.md (0 blocking, 2 MEDIUM should-fix)
- regression.md (0 blocking, 0 should-fix, 1 pre-existing LOW)
- tests.md (2 MEDIUM should-fix)
- rust.md (0 blocking, 0 should-fix, 3 pre-existing LOW)

---

**Claude Code** | Comprehensive Code Review
_Review conducted on 2026-03-15 with 80%+ confidence filtering for inline comments_
